enclave_build/src/image_manager.rs

```diff
@@ -0,0 +1,180 @@
+// Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+#![allow(dead_code)]
+
+use std::convert::TryFrom;
+
+use oci_distribution::Reference;
+
+use crate::image::ImageDetails;
+use crate::storage::OciStorage;
+use crate::{EnclaveBuildError, Result};
+
+pub struct OciImageManager {
+    /// Name of the container image.
+    image_name: String,
+    /// Image details needed for inspect and extract commands operations
+    image_details: ImageDetails,
+}
+
+impl OciImageManager {
+    /// When calling this constructor, it also tries to initialize the storage at the default path.
+    /// If this fails, the ImageManager is still created, but the 'storage' field is set to 'None'.
+    pub async fn new(image_name: &str) -> Result<Self> {
+        // Add the default ":latest" tag if the image tag is missing
+        let image_name = normalize_tag(image_name)?;
+
+        // The docker daemon is not used, so a local storage needs to be created
+        let storage =
+            match OciStorage::get_default_root_path().map_err(|err| eprintln!("{:?}", err)) {
+                Ok(root_path) => {
+                    // Try to create/read the storage. If the storage could not be created, log the error
+                    OciStorage::new(&root_path)
+                        .map_err(|err| eprintln!("{:?}", err))
+                        .ok()
+                }
+                Err(_) => None,
+            };
+
+        let image_details = Self::fetch_image_details(&image_name, storage).await?;
+
+        Ok(Self {
+            image_name,
+            image_details,
+        })
+    }
+
+    /// Returns a struct containing image metadata.
+    ///
+    /// If the image is stored correctly, the function tries to fetch the image from the storage.
+    ///
+    /// If the image is not stored or a storage was not created (the 'storage' field is None),
+    /// it pulls the image, stores it (if the 'storage' field is not None) and returns its metadata.
+    ///
+    /// If the pull succeeded but the store operation failed, it returns the pulled image metadata.
+    async fn fetch_image_details(
+        image_name: &str,
+        mut storage: Option<OciStorage>,
+    ) -> Result<ImageDetails> {
+        let local_storage = storage.as_mut();
+
+        let image_details = if let Some(storage) = local_storage {
+            // Try to fetch the image from the storage
+            storage.fetch_image_details(image_name).map_err(|err| {
+                // Log the fetching error
+                eprintln!("{:?}", err);
+                err
+            })
+        } else {
+            Err(EnclaveBuildError::OciStorageNotFound(
+                "Local storage missing".to_string(),
+            ))
+        };
+
+        // If the fetching failed, pull it from remote and store it
+        match image_details {
+            Ok(details) => Ok(details),
+            Err(_) => {
```

Comment:
I think this err might be catching a few too many things. Yes, it might be that the image is not already in storage, but it could also be that we don't have access to the storage (ie. permission issues) or that the folder doesn't exist and it might have to be created either by this library or, I think, preferably by the user.